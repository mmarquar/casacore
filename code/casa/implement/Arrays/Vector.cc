//# Vector.cc: A 1-D Specialization of the Array Class
//# Copyright (C) 1993,1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Arrays/Vector.h>
#include <aips/Arrays/ArrayError.h>
#include <aips/Utilities/Assert.h>
#include <aips/Mathematics/Math.h>
#include <aips/Utilities/Copy.h>
#include <iostream.h>

//# Implement rtti functions.
rtti_imp_mbrf_a1(Vector);

template<class T> Vector<T>::Vector()
{
    DebugAssert(ok(), ArrayError);
}


template<class T> Vector<T>::Vector(uInt Length, Int Origin)
: Array<T>(IPosition(1, Length), IPosition(1, Origin))
{
    DebugAssert(ok(), ArrayError);
}

template<class T> Vector<T>::Vector(uInt Length)
: Array<T>(IPosition(1, Length))
{
    DebugAssert(ok(), ArrayError);
}


template<class T> Vector<T>::Vector(const Block<T> &other, Int nr)
: Array<T>(IPosition(1, other.nelements()))
{
    initVector (other, nr);
    DebugAssert(ok(), ArrayError);
}

template<class T> Vector<T>::Vector(const Block<T> &other)
: Array<T>(IPosition(1, other.nelements()))
{
    initVector (other, 0);
    DebugAssert(ok(), ArrayError);
}

// Copy from the block. Copy the number of elements specified or
// the number of elements in the block if nr <= 0.
// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> void Vector<T>::initVector(const Block<T> &other, Int nr)
{
    uInt n = nr;
    if (nr <= 0) 
	n = other.nelements();
    if (n > other.nelements())
	throw(ArrayError("Vector<T>::initVector(const Block<T> &other"
				   ", Int nr) - nr > other.nelements()"));
    if (this->nelements() != n) {
	this->resize(n);
    }
    for (Int i=0; i < n; i++)
	(*data)[i] = other[i];

    return;
}

template<class T> Vector<T>::Vector(const Vector<T> &other)
: Array<T>(other)
{
    DebugAssert(ok(), ArrayError);
}

// <thrown>
//    <item> ArrayNDimError
// </thrown>
template<class T> Vector<T>::Vector(const Array<T> &other)
: Array<T>(other)
{
    if (ndim() != 1)
	throw(ArrayNDimError(1, other.ndim(), "Vector<T>::Vector"
			     " (const Array<T> &) : ndim != 1"));
    DebugAssert(ok(), ArrayError);
}

template<class T>
Vector<T>::Vector(const IPosition &shape, T *storage, 
		  StorageInitPolicy policy)
  : Array<T>(shape, storage, policy)
{
    AlwaysAssert(shape.nelements() == 1, ArrayError);
}

template<class T>
Vector<T>::Vector(const IPosition &shape, const T *storage)
  : Array<T>(shape, storage)
{
    AlwaysAssert(shape.nelements() == 1, ArrayError);
}

template<class T> Vector<T>::~Vector()
{
    // Nothing
}


// <thrown>
//    <item> ArrayConformanceError
// </thrown>
template<class T> void Vector<T>::resize(const IPosition &l, const IPosition &o,
					 Bool copyValues)
{
    DebugAssert(ok(), ArrayError);
    if (l.nelements() != 1 || o.nelements() != 1)
	throw(ArrayConformanceError("Vector<T>::resize() - attempt to form "
				    "non-vector"));
    if (copyValues) {
        Vector<T> oldref(*this);
	Array<T>::resize(l,o);
	uInt minNels = min(this->nelements(), oldref.nelements());
	objcopy(begin, oldref.begin, minNels, uInt(inc[0]), uInt(oldref.inc[0]));
    } else {
	Array<T>::resize(l,o);
    }
}

template<class T> void Vector<T>::resize(const IPosition &l, const IPosition &o)
{
    resize(l,o,False);
}

template<class T> void Vector<T>::resize(const IPosition &l)
{
    resize(l,False);
}


template<class T> void Vector<T>::resize(const IPosition &len, Bool copyValues)
{
    DebugAssert(ok(), ArrayError);
    IPosition or(len.nelements());
    or = 0;
    resize(len,or, copyValues);
}

template<class T> void Vector<T>::resize(uInt len, Int orign, Bool copyValues)
{
    DebugAssert(ok(), ArrayError);
    IPosition l(1), o(1);
    l(0) = len;
    o(0) = orign;
    Vector<T>::resize(l,o, copyValues);
}

// <thrown>
//    <item> ArrayNDimError
// </thrown>
template<class T> void Vector<T>::reference(Array<T> &other)
{
    DebugAssert(ok(), ArrayError);
    if (other.ndim() != 1)
	throw(ArrayNDimError(1, other.ndim(), "Vector<T>::reference()"
			     " - attempt to reference non-vector"));
    Array<T>::reference(other);
}

template<class T> Vector<T> &Vector<T>::operator=(const Vector<T> &other)
{
    DebugAssert(ok(), ArrayError);

    if (this == &other)
        return *this;

    Bool Conform = conform(other);
    if (Conform == False && length[0] != 0)
	validateConformance(other);  // We can't overwrite, so throw exception


    Int i;
    if (Conform == True) { // copy in place
	Int indexOffset = other.start[0] - start[0];
	for(i=start[0]; i < start[0] + length[0]; i++)
	    (*this)(i) = other(i + indexOffset);
    } else {
        data = new Block<T>(other.length[0]);
        begin = data->storage();
        start = other.start;
        length = other.length;
	nels = other.nels;
	originalLength = length;
        inc[0] = 1; // flatten
	Int indexOffset = other.start[0] - start[0];
	for(i=start[0]; i < start[0] + length[0]; i++)
	    (*this)(i) = other(i + indexOffset);
    }
    return *this;
}

// Copy a vector to a block. 
template<class T> void Vector<T>::toBlock(Block<T> & other) const
{
  DebugAssert(ok(), ArrayError);
  uInt vec_length = nelements();
// Make sure the block has enough space, but there is no need to copy elements
  other.resize(vec_length, True, False);
// Do not assume that the input Vector starts at zero.
  Int j;
  (*this).origin(j);
// Now copy the data across
  for (uInt i = 0; i < vec_length; )
    other[i++] = (*this)(j++);
}

template<class T> Array<T> &Vector<T>::operator=(const Array<T> &a)
{
    DebugAssert(ok(), ArrayError);
    Vector<T> tmp(a);
    (*this) = tmp;
    return *this;
}

// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> Vector<T> Vector<T>::operator()(const Slice &slice)
{
    DebugAssert(ok(), ArrayError);
    Int b, l, s;       // begin length step
    if (slice.all()) {
	b = start[0];
	l = length[0];
	s = 1;
    } else {
	b = slice.start();
	l = slice.length();
	s = slice.inc();
    }

    // Check that the selected slice is valid
    if (s < 1) {
	throw(ArrayError("Vector<T>::operator()(Slice) : step < 1"));
    } else if (l < 0) {
	throw(ArrayError("Vector<T>::operator()(Slice) : length < 0"));
    } else if (b+(l-1)*s > start[0]+length[0] - 1) {
	throw(ArrayError("Vector<T>::operator()(Slice) : Desired slice extends"
			 " beyond the end of the array"));
    } else if (b < start[0]) {
	throw(ArrayError("Vector<T>::operator()(Slice) : start of slice before "
			 "beginning of vector"));
    }

    // Create the slice. This could also be done with the Array<T>::operator()
    // slice functions, however it's simple for vectors, and this will be
    // more efficient.

    // Create the vector that will be the slice into this
    Vector<T> vp(*this);

    // Increment vp's begin so that it is at the selected position
    vp.begin += (b - vp.start[0])*vp.inc[0];

    vp.start[0] = 0;    // Slices always zero based
    vp.inc[0] *= s;

    vp.length[0] = l;
    vp.nels = l;

    return vp;
}

template<class T> IPosition Vector<T>::origin() const
{
    DebugAssert(ok(), ArrayError);
    return Array<T>::origin();
}

template<class T> IPosition Vector<T>::shape() const
{
    DebugAssert(ok(), ArrayError);
    return Array<T>::shape();
}

template<class T> IPosition Vector<T>::end() const
{
    DebugAssert(ok(), ArrayError);
    return Array<T>::end();
}

template<class T>
void Vector<T>::takeStorage(const IPosition &shape, T *storage,
		     StorageInitPolicy policy)
{
    AlwaysAssert(shape.nelements() == 1, ArrayError);
    Array<T>::takeStorage(shape, storage, policy);
}

template<class T>
void Vector<T>::takeStorage(const IPosition &shape, const T *storage)
{
    AlwaysAssert(shape.nelements() == 1, ArrayError);
    Array<T>::takeStorage(shape, storage);
}

template<class T> Bool Vector<T>::ok() const
{
    return ( (ndim() == 1) ? (Array<T>::ok()) : False );
}

