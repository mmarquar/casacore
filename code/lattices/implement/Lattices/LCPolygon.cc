//# LCPolygon.cc: Define a 2-dimensional region by a polygon
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <trial/Lattices/LCPolygon.h>
#include <aips/Arrays/Matrix.h>
#include <aips/Arrays/ArrayMath.h>
#include <aips/Mathematics/Math.h>
#include <aips/Tables/TableRecord.h>
#include <aips/Utilities/GenSort.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h>

typedef Vector<Int> lcpolygon_gppbug1;


LCPolygon::LCPolygon()
{}

LCPolygon::LCPolygon (const Vector<Float>& x, const Vector<Float>& y,
		      const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsX (x.copy()),
  itsY (y.copy())
{
    defineBox();
    defineMask();
}

LCPolygon::LCPolygon (const Vector<Double>& x, const Vector<Double>& y,
		      const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsX (x.nelements()),
  itsY (y.nelements())
{
    for (uInt i=0; i<x.nelements(); i++) {
	itsX(i) = x(i);
	if (i < y.nelements()) {
           // define box will catch unequal lengths x and y
           itsY(i) = y(i);  
        }
                                                
    }
    defineBox();
    defineMask();
}

LCPolygon::LCPolygon (const LCPolygon& other)
: LCRegionFixed (other),
  itsX (other.itsX),
  itsY (other.itsY)
{}

LCPolygon::~LCPolygon()
{}

LCPolygon& LCPolygon::operator= (const LCPolygon& other)
{
    if (this != &other) {
	LCRegionFixed::operator= (other);
	itsX.resize (other.itsX.nelements());
	itsY.resize (other.itsX.nelements());
	itsX = other.itsX;
	itsY = other.itsY;
    }
    return *this;
}


Bool LCPolygon::operator== (const LCRegion& other) const
{
    // Check if parent class matches.
    // If so, we can safely cast.
    if (! LCRegionFixed::operator== (other)) {
	return False;
    }
    const LCPolygon& that = (const LCPolygon&)other;
    // Compare private data.
    if (itsX.nelements() != that.itsX.nelements()
    ||  itsY.nelements() != that.itsY.nelements()) {
	return False;
    }
    Bool deleteX1, deleteY1;
    Bool deleteX2, deleteY2;
    const Float* pX1 = itsX.getStorage (deleteX1);
    const Float* pY1 = itsY.getStorage (deleteY1);
    const Float* pX2 = that.itsX.getStorage (deleteX2);
    const Float* pY2 = that.itsY.getStorage (deleteY2);
    Bool result = True;
    for (uInt i=0; i<itsX.nelements(); i++) {
	if (!near (pX1[i], pX2[i])
	||  !near (pY1[i], pY2[i])) {
	    result = False;
	    break;
	}
    }
    itsX.freeStorage (pX1, deleteX1);
    itsY.freeStorage (pY1, deleteY1);
    that.itsX.freeStorage (pX2, deleteX2);
    that.itsY.freeStorage (pY2, deleteY2);
    return result;
}


LCRegion* LCPolygon::cloneRegion() const
{
    return new LCPolygon(*this);
}

LCRegion* LCPolygon::doTranslate (const Vector<Float>& translateVector,
				  const IPosition& newLatticeShape) const
{
    Vector<Float> x, y;
    x = itsX;
    y = itsY;
    uInt n = x.nelements();
    for (uInt i=0; i<n; i++) {
        x(i) += translateVector(0);
        y(i) += translateVector(1);
    }
    return new LCPolygon (x, y, newLatticeShape);
}

String LCPolygon::className()
{
    return "LCPolygon";
}

String LCPolygon::type() const
{
   return className();
}


TableRecord LCPolygon::toRecord (const String&) const
{
    TableRecord rec;
    defineRecordFields (rec, className());
    // Write 1-relative.
    rec.define ("oneRel", True);
    rec.define ("x", itsX.ac() + Float(1));
    rec.define ("y", itsY.ac() + Float(1));
    rec.define ("shape", latticeShape().asVector());
    return rec;
}

LCPolygon* LCPolygon::fromRecord (const TableRecord& rec,
				  const String&)
{
    // If 1-relative, subtract 1 from x and y.
    Bool oneRel = rec.asBool ("oneRel");
    Float off = (oneRel ? 1:0);
    Array<Float> x (rec.asArrayFloat ("x"));
    Array<Float> y (rec.asArrayFloat ("y"));
    return new LCPolygon (x-off, y-off,
                          Vector<Int>(rec.asArrayInt ("shape")));
}

void LCPolygon::defineBox()
{
    const IPosition& shape = latticeShape();
    uInt i;
    uInt nrp = itsX.nelements();
    // First make sure basic things are right.
    if (itsY.nelements() != nrp) {
	throw (AipsError ("LCPolygon::LCPolygon - "
			  "x and y vectors must have equal length"));
    }
    if (shape.nelements() != 2) {
	throw (AipsError ("LCPolygon::LCPolygon - "
			  "can only be used as a 2-dim region"));
    }
    // If the last point is not equal to the first one, add it.
    if (!near (itsX(nrp-1), itsX(0))  ||  !near(itsY(nrp-1), itsY(0))) {
	itsX.resize (nrp+1, True);
	itsY.resize (nrp+1, True);
	nrp++;
    }
    itsX(nrp-1) = itsX(0);        // Make sure they are always equal.
    itsY(nrp-1) = itsY(0);
    if (nrp < 4) {
	throw (AipsError ("LCPolygon::LCPolygon - "
			  "at least 3 different points have to be specified"));
    }
    // Determine the maximum and minimum x,y.
    // They form the bounding box.
    // Check if at least one point is inside lattice.
    Float minx = itsX(0);
    Float maxx = itsX(0);
    Float miny = itsY(0);
    Float maxy = itsY(0);
    Bool inside = False;
    for (i=1; i<nrp; i++) {
	if (itsX(i) < minx) minx = itsX(i);
	if (itsX(i) > maxx) maxx = itsX(i);
	if (itsY(i) < miny) miny = itsY(i);
	if (itsY(i) > maxy) maxy = itsY(i);
	if (itsX(i) >= 0  &&  itsX(i) <= shape(0)-1
        &&  itsY(i) >= 0  &&  itsY(i) <= shape(1)-1) {
	    inside = True;
	}
    }
    if (!inside) {
	throw (AipsError ("LCPolygon::LCPolygon - "
			  "all x,y points are outside the lattice"));
    }
    // Determine blc and trc. Note that float to int conversion truncates,
    // so add almost 1 to get correct blc.
    IPosition blc(2, 0);
    IPosition trc(shape);
    if (minx > 0) blc(0) = Int(minx+0.5);
    if (miny > 0) blc(1) = Int(miny+0.5);
    if (maxx < shape(0)-1) trc(0) = Int(maxx+0.5);
    if (maxy < shape(1)-1) trc(1) = Int(maxy+0.5);
    setBoundingBox (Slicer(blc, trc, Slicer::endIsLast));
}

void LCPolygon::defineMask()
{
    uInt i;
    // Create and initialize the mask.
    const IPosition& blc = boundingBox().start();
    const IPosition& shape = boundingBox().length();
    Matrix<Bool> mask(shape);
    mask = False;
    uInt nrline = itsX.nelements() - 1;
    Block<Float> a(nrline);
    Block<Float> b(nrline);
    Block<Int>  dir(nrline, -1);     // -1=same dir, 0=vertical, 1=change dir
    // Fill the mask for all vertical lines.
    // Also determine the index of the last non-vertical line, which
    // is used in the slope-calculation loop.
    Int prev = -1;
    for (i=0; i<nrline; i++) {
	if (!near (itsX(i), itsX(i+1))) {
	    prev = i;
	} else {
	    dir[i] = 0;                         // vertical line
	    Int x = Int(itsX(i) + 0.5) - blc(0);
	    Int ys = Int(itsY(i)+0.5) - blc(1);
	    Int ye = Int(itsY(i+1)+0.5) - blc(1);
	    fillLine (mask, x, ys, ye);
	}
    }
    // Calculate the slope (a) and offset (b) for all line segments.
    // Vertical line segments are ignored.
    // Determine if a line segment changes direction with respect to
    // the last non-vertical line segment.
    for (i=0; i<nrline; i++) {
	if (dir[i] != 0) {
	    a[i] = (itsY(i) - itsY(i+1)) / (itsX(i) - itsX(i+1));
	    b[i] = itsY(i) - a[i] * itsX(i);
	    if ((itsX(i) > itsX(i+1)  &&  itsX(prev) < itsX(prev+1))
	    ||  (itsX(i) < itsX(i+1)  &&  itsX(prev) > itsX(prev+1))) {
		dir[i] = 1;
	    }
	    prev = i;
	}
    }
    // Loop through all x-es and determine which y-points are inside.
    // This is done by determining the crossing point of all the x-lines
    // with all line segments (the in/out algorithm).
    Block<Int> cross(nrline);
    for (Int x=0; x<shape(0); x++) {
	uInt nrcross = 0;
	Float xf = x + blc(0);
	for (i=0; i<nrline; i++) {
	    // Ignore vertical lines.
	    if (dir[i] != 0) {
		// A pixel is not a single point but in fact an interval.
		Float xfs = xf - 0.5;
		Float xfe = xf + 0.5;
		Int cr = -1;
		// When a polygon point is a pixel point (in x), always
		// count the ending point of the line segment.
		// Count the starting point only if the direction has
		// changed with respect to the previous non-vertical segment.
		if (itsX(i) > xfs  &&  itsX(i) < xfe) {
		    if (dir[i] == 1) {
			cr = Int(itsY(i) + 0.5) - blc(1);
		    }
		} else if (itsX(i+1) > xfs  &&  itsX(i+1) < xfe) {
		    cr = Int(itsY(i+1) + 0.5) - blc(1);
		} else {
		    // Calculate the crossing point for the interval around xf
		    // if that interval and the line segment overlap.
		    if (xfs < itsX(i+1)  &&  xfe > itsX(i)
		    || (xfs < itsX(i)  &&  xfe > itsX(i+1))) {
			cr = Int(a[i] * xfs + b[i] + 0.5) - blc(1);
			Int cre = Int(a[i] * xfe + b[i] + 0.5) - blc(1);
			if (cr != cre) {
			    fillLine (mask, x, cr, cre);
			}
		    }
		}
		if (cr >= 0) {
		    cross[nrcross++] = cr;
		}
	    }
	}
	DebugAssert (nrcross >= 2, AipsError);
	DebugAssert (nrcross%2 == 0, AipsError);
	GenSort<Int>::sort (cross, nrcross);
	for (i=0; i<nrcross; i+=2) {
	    Int ys = cross[i];
	    Int ye = cross[i+1];
	    while (ys <= ye) {
		mask(x,ys++) = True;
	    }
	}
    }
    setMask (mask);
}

void LCPolygon::fillLine (Matrix<Bool>& mask, Int x, Int y1, Int y2)
{
    Int ys = y1;
    if (y2 < y1) {
	ys = y2;
	y2 = y1;
    }
    while (ys <= y2) {
	mask(x,ys++) = True;
    }
}
