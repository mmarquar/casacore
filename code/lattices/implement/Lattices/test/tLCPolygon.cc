//# tLCPolygon.cc: Test program for LCPolygon class
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <trial/Lattices/LCPolygon.h>
#include <aips/Tables/TableRecord.h>
#include <aips/Arrays/Vector.h>
#include <aips/Arrays/ArrayIO.h>
#include <aips/Arrays/ArrayLogical.h>
#include <aips/Utilities/Assert.h>
#include <iostream.h>


void doIt (const IPosition& latticeShape,
	   const Vector<Float>& x,
	   const Vector<Float>& y)
{
    LCPolygon polygon (x, y, latticeShape);
    cout << polygon.hasMask() << ' ' << polygon.maskArray() << endl;
    cout << polygon.box().start() << polygon.box().end()
	 << polygon.box().length() << polygon.latticeShape() << endl;
    cout << polygon.x().ac() << polygon.y().ac() << endl;
}


main()
{
    try {
	{
	    // A simple rectangle.
	    Vector<Float> x(4), y(4);
	    x(0)=3; y(0)=3;
	    x(1)=6; y(1)=3;
	    x(2)=6; y(2)=5;
	    x(3)=3; y(3)=5;
	    doIt (IPosition (2,11,20), x, y);
	    // Make right side a bit different.
	    x(2)=8;
	    doIt (IPosition (2,11,20), x, y);
	}
	{
	    // A rectangle with bottom-right corner cut off.
	    Vector<Float> x(5), y(5);
	    x(0)=3; y(0)=3;
	    x(1)=5; y(1)=3;
	    x(2)=6; y(2)=4;
	    x(3)=6; y(3)=5;
	    x(4)=3; y(4)=5;
	    doIt (IPosition (2,11,20), x, y);
	    // Make right side a bit different.
	    x(2)=8;
	    doIt (IPosition (2,11,20), x, y);
	}
	{
	    Vector<Float> x(7), y(7);
	    x(0)=3; y(0)=3;
	    x(1)=5; y(1)=3;
	    x(2)=5; y(2)=5;
	    x(3)=2; y(3)=5;
	    x(4)=3; y(4)=4;
	    x(5)=2; y(5)=3;
	    x(6)=2; y(6)=2;
	    doIt (IPosition (2,11,20), x, y);
	    x(3)=3;
	    doIt (IPosition (2,11,20), x, y);
	}
	{
	    // A rectangle with an inner rectangle.
	    Vector<Float> x(11), y(11);
	    x(0)=3; y(0)=3;
	    x(1)=9; y(1)=3;
	    x(2)=9; y(2)=8;
	    x(3)=3; y(3)=8;
	    x(4)=3; y(4)=3;
	    x(5)=5; y(5)=5;
	    x(6)=7; y(6)=5;
	    x(7)=7; y(7)=7;
	    x(8)=5; y(8)=7;
	    x(9)=5; y(9)=5;
	    x(10)=3; y(10)=3;
	    doIt (IPosition (2,11,20), x, y);
	    x(6)=8; y(6)=4;
	    doIt (IPosition (2,11,20), x, y);
	}
	{
	    // A cross-like figure
	    Vector<Float> x(4), y(4);
	    x(0)=3; y(0)=3;
	    x(1)=9; y(1)=3;
	    x(2)=3; y(2)=8;
	    x(3)=9; y(3)=8;
	    doIt (IPosition (2,11,20), x, y);
	}
	{
	    // A pentagram (with the inner pentagon excluded)
	    Vector<Float> x(5), y(5);
	    x(0)=1; y(0)=0;
	    x(1)=3; y(1)=4;
	    x(2)=5; y(2)=0;
	    x(3)=0; y(3)=3;
	    x(4)=6; y(4)=3;
	    doIt (IPosition (2,11,20), x, y);
	}
    } catch (AipsError x) {
	cout << "Caught exception: " << x.getMesg() << endl;
	return 1;
    } end_try;
    cout << "OK" << endl;
    return 0;
}
