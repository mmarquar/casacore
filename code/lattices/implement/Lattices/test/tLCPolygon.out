1 Axis Lengths: [4, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1
 1, 1, 1]

[3, 3][6, 5][4, 3][11, 20]
[3, 6, 6, 3, 3][3, 3, 5, 5, 3]
1 Axis Lengths: [6, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1
 1, 1, 1
 0, 1, 1
 0, 0, 1]

[3, 3][8, 5][6, 3][11, 20]
[3, 6, 8, 3, 3][3, 3, 5, 5, 3]
1 Axis Lengths: [4, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1
 0, 1, 1]

[3, 3][6, 5][4, 3][11, 20]
[3, 5, 6, 6, 3, 3][3, 3, 4, 5, 5, 3]
1 Axis Lengths: [6, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1
 1, 1, 1
 0, 1, 1
 0, 1, 0]

[3, 3][8, 5][6, 3][11, 20]
[3, 5, 8, 6, 3, 3][3, 3, 4, 5, 5, 3]
1 Axis Lengths: [4, 4]  (NB: Matrix in Row/Column order)
[1, 1, 0, 1
 0, 1, 1, 1
 0, 1, 1, 1
 0, 1, 1, 1]

[2, 2][5, 5][4, 4][11, 20]
[3, 5, 5, 2, 3, 2, 2, 3][3, 3, 5, 5, 4, 3, 2, 3]
1 Axis Lengths: [4, 4]  (NB: Matrix in Row/Column order)
[1, 1, 0, 0
 0, 1, 1, 1
 0, 1, 1, 1
 0, 1, 1, 1]

[2, 2][5, 5][4, 4][11, 20]
[3, 5, 5, 3, 3, 2, 2, 3][3, 3, 5, 5, 4, 3, 2, 3]
1 Axis Lengths: [7, 6]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 0, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1]

[3, 3][9, 8][7, 6][11, 20]
[3, 9, 9, 3, 3, 5, 7, 7, 5, 5, 3][3, 3, 8, 8, 3, 5, 5, 7, 7, 5, 3]
1 Axis Lengths: [7, 6]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 0, 1, 1
 1, 1, 1, 0, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1]

[3, 3][9, 8][7, 6][11, 20]
[3, 9, 9, 3, 3, 5, 8, 7, 5, 5, 3][3, 3, 8, 8, 3, 5, 4, 7, 7, 5, 3]
1 Axis Lengths: [7, 6]  (NB: Matrix in Row/Column order)
[1, 0, 0, 0, 0, 1
 1, 1, 0, 0, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 1, 1, 1, 1
 1, 1, 0, 0, 1, 1
 1, 0, 0, 0, 0, 1]

[3, 3][9, 8][7, 6][11, 20]
[3, 9, 3, 9, 3][3, 3, 8, 8, 3]
1 Axis Lengths: [7, 5]  (NB: Matrix in Row/Column order)
[0, 0, 0, 1, 0
 1, 0, 1, 1, 0
 1, 1, 1, 1, 0
 0, 1, 1, 1, 1
 1, 1, 1, 1, 0
 1, 0, 1, 1, 0
 0, 0, 0, 1, 0]

[0, 0][6, 4][7, 5][11, 20]
[1, 3, 5, 0, 6, 1][0, 4, 0, 3, 3, 0]
OK
